providers/generic.js

```diff
@@ -1,18 +1,110 @@
+        /**
+         * Generic OAuth2 flow
+         * 
+         * 1. Obtain an access token from Google Authorization server
+         * 2. Examine scopes of access granted by user
+         * 3. Send the access token to an API
+         * 
+         */
+class GenericHandler {
 
-// {
-//     "client_id": '',
-//     "client_secret": '',
-//     "redirect_uri": '',
-//     "scopes":"",
+    constructor(configobj,extraOptions = {}) {
+        if (!configobj)
+            throw "No configuration object provided"
+            
+        //Required
+        this.client_id = configobj.client_id;
+        this.client_secret = configobj.client_secret;
+        this.redirect_uri = configobj.redirect_uri; //We recommend setting this to something like https://example.org/api/oauth2/google
+        this.auth_base_url = configobj.auth_base_url
+        this.token_url = configobj.token_url
 
-//     "authentication_uri":'',
-//     "platform_name":'',
-//     "token_uri":'',
-//     "":""
-// }
+        //Optional
+        this.response_type = configobj.response_type ? configobj.response_type : null;
+        this.scope = configobj.scope ? configobj.scope : null;//Default to profile scope if no scope is defined    -  && configobj.isArray()
+        this.platform_name = configobj.platform_name ? configobj.platform_name : 'generic';
 
-// // Query for access code
+        this.libs = {};
+        this.libs.fetch = require('node-fetch');
+        this.libs.log = require('../lib/logging-debug');
+        this.libs.checkStatus = (res) => {if (res.ok) {return res} else { console.log(res); throw "Unhealthy response"}};
+    }
 
-// // Query for access token
+    startFlow() {//Should return a uri to begin the OAuth2 flow and gain user consent
+        this.libs.log.info(`[${this.platform_name}] Start of OAuth2 flow, generating redirect uri to gain user consent`);
+        try {
+            return `${this.auth_base_url}?client_id=${this.client_id}&response_type=${this.response_type}&scope=${this.scope}&redirect_uri=${this.redirect_uri}/callback`;
+        } catch (e) {
+            this.libs.log.error(`[${this.platform_name}] Failed to start OAuth2 flow: Couldn't generate (and/or) return the consent uri`);
+            throw `[${this.platform_name}] Failed to generate consent uri\n${e}`;
+        }
+    }
 
-// // Query for data (may differ across platforms)
\ No newline at end of file
+    async stopFlow(flowResponse) {//Should receive the token, automatically and prepare it for the user - the token is not stored and this should return USER DATA only
+        return new Promise (async (resolve, reject) => {
```

Comment:
Shouldn't have an async function inside your promise, although if we do not we will never be able to fulfil the promise as we require it

Reply:
@0xh4sh

Reply:
Solved